File: dsa/12-graph/bfsGraph.py
from collections import deque
from typing import List


class Graph:
    def __init__(self, v) -> None:
        self.v = v
        self.adjList = [[0] for _ in range(v)]
        self.adjHashSets = {}

    def addEdge(self, s, d) -> None:
        self.adjList[s].append(d)
        self.adjList[d].append(s)

    def bfs(self, node: int):
        visited = set()
        visited.add(node)
        queue = deque()
        queue.append(node)

        while queue:
            curr = queue.popleft()
            print(curr, end=" ")
            for neighbor in self.adjList[curr]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

    def bfsWithHashSets(self, node: int):
        visited = set()
        visited.add(node)
        queue = deque()
        queue.append(node)

        while queue:
            curr = queue.popleft()
            print(curr, end=" ")
            for neighbor in self.adjHashSets[curr]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

    def convertEdges2AdjList(self, edges: List[List[int]]):
        for src, des in edges:
            self.adjList[src].append(des)
            self.adjList[des].append(src)

    def convertEdges2AdjListHashSet(self, edges: List[List[int]]):
        for src, des in edges:
            if src not in self.adjHashSets:
                self.adjHashSets[src] = []
            if des not in self.adjHashSets:
                self.adjHashSets[des] = []

            self.adjHashSets[src].append(des)
            self.adjHashSets[des].append(src)

    def convertMatrix2AdjList(self, matrix: List[List[int]]):
        for r, row in enumerate(matrix):
            for c, _ in enumerate(row):
                if matrix[r][c] == 1:
                    self.adjList[r].append(c)
                    self.adjList[c].append(r)

    def convertMatrix2AdjListHashSet(self, matrix: List[List[int]]):
        for r, row in enumerate(matrix):
            for c, _ in enumerate(row):
                if matrix[r][c] == 1:
                    if r not in self.adjHashSets:
                        self.adjHashSets[r] = []
                    if c not in self.adjHashSets:
                        self.adjHashSets[c] = []

                    self.adjHashSets[r].append(c)
                    self.adjHashSets[c].append(r)


if __name__ == "__main__":
    g = Graph(4)
    """ block: insert edge one at a time"""
    # g.addEdge(0, 1)
    # g.addEdge(0, 2)
    # g.addEdge(1, 2)
    # g.addEdge(2, 0)
    # g.addEdge(2, 3)
    # g.addEdge(3, 3)
    # g.bfs(2)
    """end block: insert edge one at a time"""

    edges = [[0, 1], [0, 2], [1, 2], [2, 0], [2, 3], [3, 3]]
    matrix = [[0, 1, 1, 0], [0, 0, 1, 0], [1, 0, 0, 1], [0, 0, 0, 1]]
    ### Edges to Adjacency List
    # g.convertEdges2AdjList(edges)
    # g.bfs(2)
    ### End of Edges to Adjacency List

    ### Edges with adjacency list in HashSet
    # g.convertEdges2AdjListHashSet(edges)
    # g.bfsWithHashSets(2)
    ###

    ### Matrix to Adjacency List
    g.convertMatrix2AdjList(matrix)
    g.bfs(2)
    ### End of Matrix to Adjacency List
    print()

    ### Matrix to Adjacency List in hashset
    g.convertMatrix2AdjListHashSet(matrix)
    g.bfsWithHashSets(2)
    ### Matrix to Adjacency List in hashset


"""
SC: O(V)+O(E)
SC: O(n)+O(n)=O(n)
"""

File: dsa/12-graph/bfsGraph.cpp
#include <iostream>
#include <vector>
#include <set>
#include <unordered_set>
#include <unordered_map>
#include <queue>

using namespace std;

#pragma region bfs
void addEdge(vector<int> adj[], int s, int d)
{
    adj[s].push_back(d);
    adj[d].push_back(s);
}
void bfs(int node, vector<int> adj[], unordered_map<int, bool> &visited)
{
    queue<int> queue;
    queue.push(node);
    visited[node] = 1;
    while (queue.size())
    {
        int curr = queue.front();
        queue.pop();
        cout << curr << " ";
        for (auto neighbor : adj[curr])
        {
            if (!visited[neighbor])
            {
                visited[neighbor] = 1;
                queue.push(neighbor);
            }
        }
    }
}
#pragma endregion bfs

#pragma region bfs from a given list of edges
vector<vector<int>> convertEdgesToAdjList(vector<vector<int>> edges, int v)
{
    vector<vector<int>> adj(v);
    for (auto edge : edges)
    {
        adj[edge[0]].push_back(edge[1]);
        adj[edge[1]].push_back(edge[0]);
    }
    return adj;
}

void bfsEdges2AdjList(int node, vector<vector<int>> adj)
{
    unordered_set<int> visited;
    queue<int> queue;
    queue.push(node);
    visited.insert(node);
    while (queue.size())
    {
        int curr = queue.front();
        queue.pop();
        cout << curr << " ";
        for (auto neighbor : adj[curr])
        {
            if (visited.count(neighbor) == 0)
            {
                visited.insert(neighbor);
                queue.push(neighbor);
            }
        }
    }
}
#pragma endregion bfs from a given list of edges

#pragma region bfs for given matrix
unordered_map<int, vector<int>> convertMatrix2AdjList(vector<vector<int>> matrix)
{
    int numRows = matrix.size();
    int numColums = matrix[0].size();
    unordered_map<int, vector<int>> matrix2AdjList;
    for (int row = 0; row < numRows; row++)
    {
        for (int col = 0; col < numColums; col++)
        {
            if (matrix[row][col] == 1)
            {
                matrix2AdjList[row].push_back(col);
                matrix2AdjList[col].push_back(row);
            }
        }
    }
    return matrix2AdjList;
}

void bfsMatrix2AdjList(int node, unordered_map<int, vector<int>> matrix2AdjList)
{
    unordered_set<int> visited;
    queue<int> queue;
    queue.push(node);
    visited.insert(node);
    while (queue.size())
    {
        int curr = queue.front();
        queue.pop();
        cout << curr << " ";
        for (auto neighbor : matrix2AdjList[curr])
        {
            if (visited.count(neighbor) == 0)
            {
                visited.insert(neighbor);
                queue.push(neighbor);
            }
        }
    }
}
#pragma endregion bfs from a given matrix
int main()
{
    int v = 4;
    vector<int> adj[v];
    unordered_map<int, bool> visited;
    addEdge(adj, 0, 1);
    addEdge(adj, 0, 2);
    addEdge(adj, 1, 2);
    addEdge(adj, 2, 0);
    addEdge(adj, 2, 3);
    addEdge(adj, 3, 3);
    bfs(2, adj, visited);
    cout << endl;

    vector<vector<int>> edges = {{0, 1}, {0, 2}, {1, 2}, {2, 0}, {2, 3}, {3, 3}};
    vector<vector<int>> edges2AdjList = convertEdgesToAdjList(edges, v);
    bfsEdges2AdjList(2, edges2AdjList);

    cout << endl;
    vector<vector<int>> matrix = {
        {0, 1, 1, 0},
        {0, 0, 1, 0},
        {1, 0, 0, 1},
        {0, 0, 0, 1}};
    unordered_map<int, vector<int>> matrix2AdjList = convertMatrix2AdjList(matrix);
    bfsMatrix2AdjList(2, matrix2AdjList);

    return 0;
}

/*
SC: O(V)+O(2E)=O(V+E)
SC: O(n)+O(n)=O(n)
*/